File: NumMeth/Demos/float-to-ascii/fpout.c
//
//

//* fpout.C
//* C Version of the floating point to ASCII conversion function.
//* (C) 2014 DJ Greaves, University of Cambridge, Computer Laboratory.
//*


#include <stdio.h>

int tens_table[]  = { 1,10,100,1000,10000,100000 };



void bin2ascii(unsigned int d0)
{
 
  // let fun scanup p = if Array.sub(tens_table, p) > d0 then p-1 else scanup(p+1)
  int p0 = 0;
  while (tens_table[p0] <= d0) p0 ++;
  p0 --;

   int ptr = 0;
   //fun digits d0 p =
   //            if p<0 then [] else
   //            let val d = d0 div Array.sub(tens_table, p)
   //                val r = d0 - d * Array.sub(tens_table, p)
   //            in chr(48 + d) :: digits r (p-1) end
   //   in digits d0 p0 end

   while(p0>=0) // Exercise: fix the printing of the case that does not work!
   {
     //printf("p0=%i d0=%i\n", p0, d0);
     int d = d0 / tens_table[p0];
     d0 = d0 - d * tens_table[p0];
     putchar((char)d + '0');
     p0 --;
   }
}



const int N_BIN_POWERS_OF_TEN = 6; // Range: 10**(+/- 63)


int dec32_table[] = { 1,10,100,1000,10000,100000 };

const long double bin_exponents_table[] =        { 1.0e1, 1.0e2, 1.0e4, 1.0e8, 1.0e16, 1.0e32 };

const long double bin_fracs_table[] =            { 1.0e-1, 1.0e-2, 1.0e-4, 1.0e-8, 1.0e-16, 1.0e-32 };

// The C version is a little richer and a little longer: it prints NaN or Inf as described in the next section of these notes.

void doubleToStr(int scientific_form, double d0, int precision)
{
  if (d0 < 0.0)
    { d0 = -d0;
      putchar('-');
    }
  union tobits
  { // Abuse union safety to get bits.
    double df;
    unsigned long long int di;
  } to;
  to.df = d0;
  int in_exp = (to.di >> 52) & 0x7FF;
  int mant = to.di & ((1llu<<52)-1);
  if (in_exp == 0x7FF)
    {
      if (mant == 0)
	{ putchar('I');          putchar('n');           putchar('f');
        }
      else
        { putchar('N');          putchar('a');           putchar('N');
        }
      return;
    }
  // Render all denormals as 0.0
  if (in_exp == 0)
    { putchar('0');
      putchar('.');
      putchar('0');
      return;
    }

  int exponent = 0;
  unsigned int i, bitpos = 1U<<(N_BIN_POWERS_OF_TEN-1);
  int lower_bound = tens_table[precision];
  int upper_bound = tens_table[precision+1];
  if (d0 < lower_bound)
    { for(i=N_BIN_POWERS_OF_TEN-1, 
	    bitpos = 1U<<(N_BIN_POWERS_OF_TEN-1);
	  bitpos; i--, bitpos>>=1)
	{ double q;
	  q = d0 * bin_exponents_table[i];
        if (q < upper_bound) 
          {
            d0 = q;
            exponent -= bitpos;
          }
	}
    }
  else
    { for(i=N_BIN_POWERS_OF_TEN-1,
	    bitpos = 1U<<(N_BIN_POWERS_OF_TEN-1);
	  bitpos; i--, bitpos>>=1)
	{ double q;
	  q = d0 * bin_fracs_table[i];
	  if (q >= lower_bound)
	    { d0 = q;
	      exponent += bitpos;
	    }
	}
    }
  exponent += precision;

  int imant = (int)d0;
  
  // Decimal point will only move a certain distance:
  // outside that range force scientific form.
  if (exponent> precision || exponent <0)
    scientific_form = 1;  
  
  int p, enable_trailzero_supress = 0;
  for (p = precision; p >= 0; p--)
    { int d = 0;
      while (imant>=dec32_table[p])
        { imant -= dec32_table[p];
          d++;
        }
      putchar(d + '0');
      if (enable_trailzero_supress && imant == 0) break;
      if (p == precision + (scientific_form ? 0: -exponent)) 
        { putchar('.'); // Print decimal point.
          enable_trailzero_supress = 1;
        }
    }
  if (scientific_form) // Print the exponent
    { putchar('e');
      bin2ascii(exponent);
    }
}



int main()
{
  doubleToStr(1, 13.45678, 2); printf("\n");

  doubleToStr(0, 13.45678, 3); printf("\n");
  return 0;
}

// eof

File: NumMeth/Demos/float-to-ascii/fpout.ml
(* fpout.ml
 * ML Version of the floating point to ASCII conversion function.
 * (C) 2014 DJ Greaves, University of Cambridge, Computer Laboratory.
 *)

(* Fixed point integer output 0 as needed for the exponent rendering *)
   val tens_table = Array.fromList [1,10,100,1000,10000,100000,1000000,10000000, 100000000];

   fun bin2ascii d0 =
       let fun b2a d0 =
	   let fun scanup p = if Array.sub(tens_table, p) > d0 then p-1 else scanup(p+1)
	       val p0 = scanup 0
	       fun digits d0 p =
		   if p<0 then [] else
		   let val d = d0 div Array.sub(tens_table, p)
		       val r = d0 - d * Array.sub(tens_table, p)
		   in chr(48 + d) :: digits r (p-1) end
               in digits d0 p0 end
       in if (d0 < 0) then #"~" :: b2a(0-d0) else b2a d0 
       end  


  (* These tables need to be a bit bigger in reality, to handle the whole range of 16sf and E308 found in the IEEE standard... *)

  val f_tens_table = Vector.fromList [1E0,1E1,1E2,1E3,1E4,1E5,1E6,1E7,1E8];  

  val bin_exps_table =   [ (1.0E32, 32),  (1.0E16, 16),  (1.0E8, 8),  (1.0E4, 4),  (1.0E2, 2),  (1.0E1, 1) ];  
  val bin_fracs_table =  [ (1.0E~32, 32), (1.0E~16, 16), (1.0E~8, 8), (1.0E~4, 4), (1.0E~2, 2), (1.0E~1, 1) ];  



fun float_to_string precision d00 = (*Entry point*)
  let val lower_bound = Vector.sub(f_tens_table, precision)
      val upper_bound = Vector.sub(f_tens_table, precision+1)
      val (d0, sign) = if d00 < 0.0 then (0.0-d00, [#"-"]) else (d00, [])
      fun chop_upwards ((ratio, bitpos), (d0, exponent)) = 
           let val q = d0 * ratio
           in if q < upper_bound then  (q, exponent - bitpos) else (d0, exponent)
           end

      fun chop_downwards ((ratio, bitpos), (d0, exponent)) = 
           let val q = d0 * ratio
           in if q > lower_bound then  (q, exponent + bitpos) else (d0, exponent)
           end

      val (d0, exponent) = if d0 < lower_bound then foldl chop_upwards (d0, 0) bin_exps_table
                                               else foldl chop_downwards (d0, 0) bin_fracs_table

      val imant = floor d0 (* Convert mantissa to integer. *)

      val exponent = exponent + precision

      (* Decimal point will only move a certain distance: outside that range force scientific form. *)
      val scientific_form = exponent > precision orelse exponent < 0

      fun digits d0 p trailzero_supress =
           if p<0 orelse (trailzero_supress andalso d0=0) then [] else
           let val d = d0 div Array.sub(tens_table, p)
               val r = d0 - d * Array.sub(tens_table, p)
               val dot_time =  (p = precision + (if scientific_form then 0 else 0-exponent))
               val rest = digits r (p-1) (trailzero_supress orelse dot_time)
               val rest = if dot_time then #"."::rest else rest (* render decimal point *)
               in if d>9 then #"?" :: bin2ascii d0 @ #"!" :: rest else chr(ord(#"0") + d) :: rest end


      val mantissa = digits imant precision false
      val exponent = if scientific_form then #"e" :: bin2ascii exponent else []
      in (d00, imant, implode(sign @ mantissa @ exponent) )
      end
;

(* Some test data ...*)
map (float_to_string 4) [ 1.0, 10.0, 1.23456789, ~2.3E19, 4.5E~19 ];

(* > val it =
 * [(1.0, 10000, "1."),
 * (10.0, 10000, "10."),
 * (1.23456789, 12345, "1.2345"), 
 * (~2.3E19, 23000, "-2.3e19"),
 * (4.5E~19, 45000, "4.5e~19")]
 *)

(* eof *)
